### PR TITLE
IGNITE-9272 j.u.zip.CRC32algo instead of PureJavaCrc32

### PR DESCRIPTION
**Pull request by zstan**

### CODE REVIEW

**Comment by nizhikov** [comment]
Hello, @zstan 

Thank you for the PR

1. Please, resolve merge conflicts.

2. I've added some comment, most of them are minor. Please, take a look.

3. I propose to add 1 major test - the one that will tests output equality of old and new CRC calculation algorithm

4. Can you attach to ticket results of benchmarking:
  4.1 Comparsion of CRC calculation algorithms: Please, add JMH benchmark in your PR.
  4.2 Ignite regular yardstick tests with and without your improvement.

File: modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            context.results[i] = PureJavaCrc32.calcCrc32(bb, BUF_LEN);
+        }
+
+        return context.results;
+    }
+
+    /** java.util.zip.CRC32 test. */
+    @Benchmark
+    //@CompilerControl(CompilerControl.Mode.DONT_INLINE) //makes looking at assembly easier
+    public int[] Crc32(Context context) {
```
Comment (nizhikov) [inline modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java]: method name should start with lower case letter.

File: modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        /** */
+        final int[] results = new int[SIZE];
+
+        /** */
+        final byte bb[] = new byte[BUF_LEN];
+
+        /** Setup */
+        @Setup
+        public void setup() {
+            for (int ii = 0; ii < BUF_LEN; ++ii)
```
Comment (nizhikov) [inline modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java]: Can we use here just `i` variable?

File: modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                bb[ii] = (byte)ii;
```
Comment (nizhikov) [inline modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java]: 1. You should setup random byte array.
2. You should wrap it to `ByteBuffer` in `setup` method so benchmark becomes more clear.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java [hunk 4]
```diff
@@ -565,14 +566,16 @@ private void reinit(FileIO fileIO) throws IOException {
                     if (calculateCrc && !skipCrc) {
                         assert PageIO.getCrc(pageBuf) == 0 : U.hexLong(pageId);
 
-                        PageIO.setCrc(pageBuf, calcCrc32(pageBuf, pageSize));
+                        PageIO.setCrc(pageBuf, U.calcCrc(pageBuf, pageSize));
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java]: If method contract is the same, why this change?

Comment (zstan) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java]: calcCrc32 i fully replace this method.

File: modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import static org.openjdk.jmh.annotations.Scope.Thread;
+
+/**
+ * Crc benchmark.
+ */
+@State(Thread)
+@OutputTimeUnit(NANOSECONDS)
+@BenchmarkMode(AverageTime)
+@Fork(value = 1, jvmArgsAppend = {
+        //"-XX:+UseSuperWord",
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java]: done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java [hunk 6]
```diff
@@ -565,14 +566,16 @@ private void reinit(FileIO fileIO) throws IOException {
...
+                        pageBuf.position(0);
                     }
 
                     // Check whether crc was calculated somewhere above the stack if it is forcibly skipped.
-                    assert skipCrc || PageIO.getCrc(pageBuf) != 0 || calcCrc32(pageBuf, pageSize) == 0 :
+                    assert skipCrc || PageIO.getCrc(pageBuf) != 0 || U.calcCrc(pageBuf, pageSize) == 0 :
                         "CRC hasn't been calculated, crc=0";
 
-                    assert pageBuf.position() == 0 : pageBuf.position();
+                    pageBuf.position(0);
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java]: Yes, I want to fully preserve existing method behaviour.
Whenever we think it right or wrong.

The goal of this task is *`replace`* existing method.
We can discuss and refactoring in another ticket.

[the remark this answers is not in the document]
Comment (zstan) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java]: all done, TC rerun.

**Comment by nizhikov** [comment]
Please, resolve conflicts with master.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java [hunk 7]
```diff
@@ -27,6 +27,8 @@
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.zip.CRC32;
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/FilePageStore.java]: Please, resolve this.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/FileWriteAheadLogManager.java [hunk 8]
```diff
@@ -60,6 +60,7 @@
 import java.util.concurrent.locks.ReentrantLock;
 import java.util.regex.Pattern;
 import java.util.stream.Stream;
+import java.util.zip.CRC32;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/FileWriteAheadLogManager.java]: Unused import.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/FsyncModeFileWriteAheadLogManager.java [hunk 9]
```diff
@@ -56,6 +56,7 @@
 import java.util.concurrent.locks.ReentrantLock;
 import java.util.regex.Pattern;
 import java.util.stream.Stream;
+import java.util.zip.CRC32;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/FsyncModeFileWriteAheadLogManager.java]: Unused import

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 10]
```diff
@@ -10579,6 +10584,53 @@ public static String toHexString(ByteBuffer buf) {
...
+
+    /**
+     * @param crcAlgo CRC algorithm.
+     * @param buf Input buffer.
+     * @param len Buffer length.
+     * @param reset Reset crc flag.
+     *
+     * @return Crc checksum.
+     */
+    public static int calcCrcEx(CRC32 crcAlgo, ByteBuffer buf, int len, boolean reset) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java]: Looks like this method not used in the project.
If yes, let's remove it.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/db/wal/crc/IgnitePureJavaCrcCompatibility.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * PureJavaCrc32 previous crc algo realization vs java.util.zip.crc32 test.
+ */
+public class IgnitePureJavaCrcCompatibility extends TestCase {
+    /** {@inheritDoc} */
+    @Override protected void setUp() throws Exception {
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/db/wal/crc/IgnitePureJavaCrcCompatibility.java]: Why we need this method here? Let's remove it

File: modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+            context.bb.rewind();
+
+            context.results[i] = PureJavaCrc32.calcCrc32(context.bb, BUF_LEN);
+        }
+
+        return context.results;
+    }
+
+    @Benchmark
+    public int[] crc32(Context context)
```
Comment (nizhikov) [inline modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java]: formatting error.

File: modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+@Warmup(iterations = 5)
+@Measurement(iterations = 5)
+public class BenchmarkCRC
+{
+    static final int SIZE = 1024;
+    static final int BUF_LEN = 4096;
+
+    @State(Thread)
+    public static class Context
+    {
```
Comment (nizhikov) [inline modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java]: formatting error.

File: modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+@State(Thread)
+@OutputTimeUnit(NANOSECONDS)
+@BenchmarkMode(AverageTime)
+@Fork(value = 1, jvmArgsAppend = {
+        "-XX:+UnlockDiagnosticVMOptions"})
+@Warmup(iterations = 5)
+@Measurement(iterations = 5)
+public class BenchmarkCRC
+{
```
Comment (nizhikov) [inline modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java]: formatting error.

File: modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Benchmark
+    public int[] crc32(Context context)
+    {
+        for (int i = 0; i < SIZE; i++) {
+            context.bb.rewind();
+
+            context.results[i] = U.calcCrc(context.bb, BUF_LEN);
+        }
+        return context.results;
```
Comment (nizhikov) [inline modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java]: empty line required.

File: modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+@OutputTimeUnit(NANOSECONDS)
+@BenchmarkMode(AverageTime)
+@Fork(value = 1, jvmArgsAppend = {
+        "-XX:+UnlockDiagnosticVMOptions"})
+@Warmup(iterations = 5)
+@Measurement(iterations = 5)
+public class BenchmarkCRC
+{
+    static final int SIZE = 1024;
+    static final int BUF_LEN = 4096;
```
Comment (nizhikov) [inline modules/benchmarks/src/main/java/org/apache/ignite/internal/benchmarks/jmh/algo/BenchmarkCRC.java]: empty line required.
